[GTK][WPE] Make it possible to create a Damage with a maximum numbers of rectangles before uniting

carlosgcampos opened the pull request:
#### 0ecb8f66ead0dbb7dd4bf9af92e1722202f6280e
<pre>
[GTK][WPE] Make it possible to create a Damage with a maximum numbers of rectangles before uniting
<a href="https://bugs.webkit.org/show_bug.cgi?id=291073">https://bugs.webkit.org/show_bug.cgi?id=291073</a>

Reviewed by Alejandro G. Castro.

This is what we really want when tracking dirty rects in GraphicsLayer.
In that case, instead of using a fixed cell size, we compute the grid
layout based on the damage rectangle ratio, and use a different cell
size for the given grid. Create the Damage in GraphicsLayer with the
layer size but passing 32 as the maximum number of rectangles.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::Damage):
(WebCore::Damage::rectsForPainting const):
(WebCore::Damage::makeFull):
(WebCore::Damage::gridSize const):
(WebCore::Damage::initialize):
(WebCore::Damage::cellIndexForRect const):
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
(WebCore::GraphicsLayerCoordinated::setNeedsDisplayInRect):
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, MaxRectangles)):

Canonical link: <a href="https://commits.webkit.org/293355@main">https://commits.webkit.org/293355@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8126ad4d1930381b3662dd44ded794e68630e4e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18283 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8520 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103774 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49238 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100694 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18576 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26734 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75101 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32251 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101654 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14100 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89094 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55458 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13882 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7059 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48620 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83835 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7137 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106146 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25740 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18763 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84074 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26117 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85295 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83559 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28203 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5885 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19450 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15997 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25698 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30880 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25516 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28836 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27091 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e10758f16798dc6392da5502d67059474ff714f0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43632

spenap commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -142,7 +142,7 @@ class Damage {
         m_rects.clear();
         m_shouldUnite = false;
-        initialize();
+        initialize(0);
```
I would suggest two named constants: `NoRectangles` (0) and `DefaultMax`  (ideally a better name, though) (32)

spenap replied:
all the more reason to name them, then :-)

spenap commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -263,11 +263,35 @@ class Damage {
-    void initialize()
+    IntSize gridSize(int maxRectangles) const
+    {
+        const float widthToHeightRatio = static_cast<float>(m_rect.width()) / static_cast<float>(m_rect.height());
```
you can cast either width or height and make the expression a bit less verbose. I suggest the numerator as that's what you're already doing below

```suggestion
        const float widthToHeightRatio = static_cast<float>(m_rect.width()) / m_rect.height();
```

spenap commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -263,11 +263,35 @@ class Damage {
-            m_gridCells = IntSize(std::ceil(static_cast<float>(m_rect.width()) / s_cellSize), std::ceil(static_cast<float>(m_rect.height()) / s_cellSize)).expandedTo({ 1, 1 });
+            if (maxRectangles) {
+                m_gridCells = gridSize(maxRectangles);
+                m_cellSize = IntSize(std::ceil(static_cast<float>(m_rect.width()) / m_gridCells.width()), std::ceil(static_cast<float>(m_rect.height()) / m_gridCells.height()));
```
The expression we use to calculate `m_cellSize` or `m_gridCells` below is the same: an IntSize out of the ceil of operands.

There will be a constexpr ceil from C++-23, so we could extract this as a helper, `IntSize toCeiledSize(const IntRect&, const IntSize&)`. This could be inliend for now, and eventually constexpr.

carlosgcampos replied:
There's already ceiledIntSize, but we can't use that because it uses LayoutUnit that uses int as its value. Maybe we can just add ceiledIntSize that receives and IntSize, but not in this patch.

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -120,7 +120,7 @@ class Damage {
         for (int row = 0; row < m_gridCells.height(); ++row) {
             for (int col = 0; col < m_gridCells.width(); ++col) {
-                const IntRect cellRect = { m_rect.x() + col * s_cellSize, m_rect.y() + row * s_cellSize, s_cellSize, s_cellSize };
+                const IntRect cellRect = { m_rect.x() + col * m_cellSize.width(), m_rect.y() + row * m_cellSize.height(), m_cellSize.width(), m_cellSize.height() };
```
as `m_cellSize` is `IntSize` perhaps we can try:
```cpp
const IntRect cellRect = { { m_rect.x() + col * m_cellSize.width(), m_rect.y() + row * m_cellSize.height() }, m_cellSize };
```

Scony commented on Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
```diff
@@ -111,11 +111,8 @@ void GraphicsLayerCoordinated::setNeedsDisplayInRect(const FloatRect& initialRec
-        m_dirtyRegion = Damage(m_size.constrainedBetween(minDamageSize, viewVisibleSize));
-    }
+    if (!m_dirtyRegion)
+        m_dirtyRegion = Damage(m_size, Damage::Mode::Rectangles, 32);
```
I'm not sure about 32 here - doesn't this mean that for 32x32 layer we get `Damage` with up to 32 rects?

carlosgcampos replied:
This is what we had before https://commits.webkit.org/271045@main and also what Apple does. So this is basically a revert of that patch but using Damage so that we unite in a better way instead of always in the first rectangle.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f8126ad4d1930381b3662dd44ded794e68630e4e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43632

carlosgcampos committed:
[GTK][WPE] Make it possible to create a Damage with a maximum numbers of rectangles before uniting
https://bugs.webkit.org/show_bug.cgi?id=291073

Reviewed by Alejandro G. Castro.

This is what we really want when tracking dirty rects in GraphicsLayer.
In that case, instead of using a fixed cell size, we compute the grid
layout based on the damage rectangle ratio, and use a different cell
size for the given grid. Create the Damage in GraphicsLayer with the
layer size but passing 32 as the maximum number of rectangles.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::Damage):
(WebCore::Damage::rectsForPainting const):
(WebCore::Damage::makeFull):
(WebCore::Damage::gridSize const):
(WebCore::Damage::initialize):
(WebCore::Damage::cellIndexForRect const):
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
(WebCore::GraphicsLayerCoordinated::setNeedsDisplayInRect):
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, MaxRectangles)):

Canonical link: https://commits.webkit.org/293355@main

webkit-commit-queue commented:
Committed 293355@main (0ecb8f66ead0): <https://commits.webkit.org/293355@main>

Reviewed commits have been landed. Closing PR #43632 and removing active labels.
